File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> nums1(nums.begin() + 1, nums.end());
        vector<int> nums2(nums.begin(), nums.end() - 1);
        return max(nums[0], max(helper(nums1), helper(nums2)));
    }

    int helper(vector<int>& nums) {
        int r1 = 0, r2 = 0;
        for (int i = 0; i < nums.size(); i++) {
            i = i % nums.size();
            int tmp = max(nums[i] + r1, r2);
            r1 = r2;
            r2 = tmp;
        }
        return r2;
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    static int robs(int[] a, int l, int r) {
        int n = a.length;
        int prev = a[l];
        int prev2 = 0;

        for (int i = 1; i < n - 1; i++) {

            int t = a[l + i];
            if (i > 1)
                t += prev2;
            int nt = prev;
            int cur = Math.max(t, nt);
            prev2 = prev;
            prev = cur;
        }
        return prev;
    }

    public int rob(int[] a) {
        int n = a.length;
        if (n == 1)
            return a[0];
        return Math.max(robs(a, 1, n - 1), robs(a, 0, n - 2));
    }
}